# Santander Kaggle Competition

## Summary

We're going to learn data science by participating in a machine learning competition.

### Competition Description

From frontline support teams to C-suites, customer satisfaction is a key measure of success. Unhappy customers don't stick around. What's more, unhappy customers rarely voice their dissatisfaction before leaving.

Santander Bank is asking Kagglers to help them identify dissatisfied customers early in their relationship. Doing so would allow Santander to take proactive steps to improve a customer's happiness before it's too late.

In this competition, you'll work with hundreds of anonymized features to predict if a customer is satisfied or dissatisfied with their banking experience.

### Get the Data
1. Sign up for Kaggle
2. Download the data https://www.kaggle.com/c/santander-customer-satisfaction/data
3. Open the files and place the 'train' and 'test' csv files in the same folder as this notebook.



### Get our libraries

In [95]:
%matplotlib inline
# inline is an option to make our plots appear inline (instead of pop out)

import matplotlib.pyplot as plt  # our standard plotting library; see 'seaborn' as alternative
import numpy as np  # fast arrays made for scientific computing, needed for sklearn
import scipy  # scientific computing tools, needed for sklearn
import pandas as pd  # great for data manipulation, looking at data
import sklearn  # aka sci-kit learn; python machine learning, built on top of numpy, scipy

In [96]:
# Set some display options so we can see everything (since this dataset isn't too large) 
pd.options.display.max_seq_items = 500
pd.options.display.max_columns = 350
pd.options.display.precision = 3

### Glimpse at our data

We want to load our data into a pandas dataframe. You're probably familiar with one-dimensional data like lists/arrays where it's a row of data. A dataframe is two-dimensional in that there's both rows and columns; this is what an Excel sheet looks like

In [97]:
# Let's load our data
DATA_LOCATION = '/Users/williamliu/Desktop/Santander/'
df_train = pd.read_csv(DATA_LOCATION + 'train.csv')

In [98]:
df_train.head()

ID  var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
0   1     2     23                   0                        0   
1   3     2     34                   0                        0   
2   4     2     23                   0                        0   
3   8     2     37                   0                      195   
4  10     2     39                   0                        0   

   imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
0                        0                        0                        0   
1                        0                        0                        0   
2                        0                        0                        0   
3                      195                        0                        0   
4                        0                        0                        0   

   imp_op_var40_efect_ult1  imp_op_var40_efect_ult3  imp_op_var40_ult1  \
0                        0                        0                  0   
1                        0                        0                  0   
2                        0                        0                  0   
3                        0                        0                  0   
4                        0                        0                  0   

   imp_op_var41_comer_ult1  imp_op_var41_comer_ult3  imp_op_var41_efect_ult1  \
0                        0                        0                        0   
1                        0                        0                        0   
2                        0                        0                        0   
3                      195                      195                        0   
4                        0                        0                        0   

   imp_op_var41_efect_ult3  imp_op_var41_ult1  imp_op_var39_efect_ult1  \
0                        0                  0                        0   
1                        0                  0                        0   
2                        0                  0                        0   
3                        0                195                        0   
4                        0                  0                        0   

   imp_op_var39_efect_ult3  imp_op_var39_ult1  imp_sal_var16_ult1  ind_var1_0  \
0                        0                  0                   0           0   
1                        0                  0                   0           0   
2                        0                  0                   0           0   
3                        0                195                   0           0   
4                        0                  0                   0           0   

   ind_var1  ind_var2_0  ind_var2  ind_var5_0  ind_var5  ind_var6_0  ind_var6  \
0         0           0         0           1         0           0         0   
1         0           0         0           1         0           0         0   
2         0           0         0           1         1           0         0   
3         0           0         0           1         1           0         0   
4         0           0         0           1         0           0         0   

   ind_var8_0  ind_var8  ind_var12_0  ind_var12  ind_var13_0  \
0           0         0            0          0            0   
1           0         0            0          0            1   
2           0         0            0          0            0   
3           0         0            0          0            0   
4           0         0            1          1            0   

   ind_var13_corto_0  ind_var13_corto  ind_var13_largo_0  ind_var13_largo  \
0                  0                0                  0                0   
1                  1                1                  0                0   
2                  0                0                  0                0   
3                  0                0                  0                0   
4   

In [99]:
np.unique(df_train.dtypes)  # Interesting, we only have numbers (int, float) and no categories (e.g. no text, categoricals)

array([dtype('int64'), dtype('float64')], dtype=object)

In [100]:
df_train.describe()

ID       var3     var15  imp_ent_var16_ult1  \
count   76020.00   76020.00  76020.00            76020.00   
mean    75964.05   -1523.20     33.21               86.21   
std     43781.95   39033.46     12.96             1614.76   
min         1.00 -999999.00      5.00                0.00   
25%     38104.75       2.00     23.00                0.00   
50%     76043.00       2.00     28.00                0.00   
75%    113748.75       2.00     40.00                0.00   
max    151838.00     238.00    105.00           210000.00   

       imp_op_var39_comer_ult1  imp_op_var39_comer_ult3  \
count                 76020.00                 76020.00   
mean                     72.36                   119.53   
std                     339.32                   546.27   
min                       0.00                     0.00   
25%                       0.00                     0.00   
50%                       0.00                     0.00   
75%                       0.00                     0.00   
max                   12888.03                 21024.81   

       imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
count                 76020.00                 76020.00   
mean                      3.56                     6.47   
std                      93.16                   153.74   
min                       0.00                     0.00   
25%                       0.00                     0.00   
50%                       0.00                     0.00   
75%                       0.00                     0.00   
max                    8237.82                 11073.57   

       imp_op_var40_efect_ult1  imp_op_var40_efect_ult3  imp_op_var40_ult1  \
count                 76020.00                 76020.00           76020.00   
mean                      0.41                     0.57               3.16   
std                      30.60                    36.51              95.27   
min                       0.00                     0.00               0.00   
25%                       0.00                     0.00               0.00   
50%                       0.00                     0.00               0.00   
75%                       0.00                     0.00               0.00   
max                    6600.00                  6600.00            8237.82   

       imp_op_var41_comer_ult1  imp_op_var41_comer_ult3  \
count                 76020.00                 76020.00   
mean                     68.80                   113.06   
std                     319.61                   512.15   
min                       0.00                     0.00   
25%                       0.00                     0.00   
50%                       0.00                     0.00   
75%                       0.00                     0.00   
max                   12888.03                 16566.81   

       imp_op_var41_efect_ult1  imp_op_var41_efect_ult3  imp_op_var41_ult1  \
count                 76020.00                 76020.00           76020.00   
mean                     68.21                   113.23             137.24   
std                     531.90                   950.09             697.71   
min                       0.00                     0.00               0.00   
25%                       0.00                     0.00               0.00   
50%                       0.00                     0.00               0.00   
75%                       0.00                     0.00               0.00   
max                   45990.00                131100.00           47598.09   

       imp_op_var39_efect_ult1  imp_op_var39_efect_ult3  imp_op_var39_ult1  \
count                 76020.00                 76020.00           76020.00   
mean                     68.62                   113.79             140.40   
std                     535.47                   953.58             712.77   
min                       0.00                     0.00               0.00   
25%                       0.00                     0.00               0.

So this summary tells us quite a bit about the numerical part of the data. Let's break it down:

* We're training our dataset to __TARGET__, which can be 0 or 1. When we look at this column, we see that the 25%, 50%, and 75% are all 0's. This means that our dataset is imbalanced (with a bit more people being 0's than 1's.
* We take that above idea of imbalanced data and look at the rest of our variables (e.g. __imp_ent_var16_ult1__). Most of the data is 0's, but then we get a max value of 210,000.
* I think we'll need to clean the data. For example, column __var3__ has a min of -999999 and a max of 238. My guess is that -999999 stands for a state instead of there being many unique values from -999999 to 0.
* Yeah, we'll need to do some data cleaning; some numeric variables (e.g. __ind_var27_0__, __ind_var_28_0__) are just 0 for min and max. We can drop these since they're not telling us anything. We can look at the standard deviation and if it's 0, then we can get rid of because there's no change (e.g. min=70 with max=70 tells us no useful info)
* It looks like some columns are duplicates. Without even needing to sort, we can see that some columns are duplicates (e.g. __num_var13_medio_0__, __num_var13_medio__ and __ind_var34_0__, __ind_var_34__) where they have have the exact same min, max, standard deviation, mean. This will cause multicollinearity, a fancy way of saying don't have a variable that can linearly predict another variable (e.g. I'm alive and I'm dead, we don't need both since one will tell us the other).

TLDR, we need to:

* [ ] Drop duplicate columns
* [ ] Drop values that are not telling us anything (i.e. standard deviation of 0)
* [ ] Possibly clean data (e.g. -999,999)
* [ ] Compensate for unbalanced dataset (lot more TARGET of 0's than 1's)


In [101]:
# Data Cleaning - Drop values that are not telling us anything (i.e. standard deviation of 1)
print "Before Data Cleaning, data shape is: ", df_train.shape

for col in df_train:
    if df_train[col].std()==0:
        print "Dropping: ", col
        del df_train[col]

Before Data Cleaning, data shape is:  (76020, 371)
Dropping:  ind_var2_0
Dropping:  ind_var2
Dropping:  ind_var27_0
Dropping:  ind_var28_0
Dropping:  ind_var28
Dropping:  ind_var27
Dropping:  ind_var41
Dropping:  ind_var46_0
Dropping:  ind_var46
Dropping:  num_var27_0
Dropping:  num_var28_0
Dropping:  num_var28
Dropping:  num_var27
Dropping:  num_var41
Dropping:  num_var46_0
Dropping:  num_var46
Dropping:  saldo_var28
Dropping:  saldo_var27
Dropping:  saldo_var41
Dropping:  saldo_var46
Dropping:  imp_amort_var18_hace3
Dropping:  imp_amort_var34_hace3
Dropping:  imp_reemb_var13_hace3
Dropping:  imp_reemb_var33_hace3
Dropping:  imp_trasp_var17_out_hace3
Dropping:  imp_trasp_var33_out_hace3
Dropping:  num_var2_0_ult1
Dropping:  num_var2_ult1
Dropping:  num_reemb_var13_hace3
Dropping:  num_reemb_var33_hace3
Dropping:  num_trasp_var17_out_hace3
Dropping:  num_trasp_var33_out_hace3
Dropping:  saldo_var2_ult1
Dropping:  saldo_medio_var13_medio_hace3


In [102]:
print "After data cleaning, data shape is ", df_train.shape

After data cleaning, data shape is  (76020, 337)


In [110]:
# Data Cleaning - Look at duplicate columns
print df_train['num_var13_medio_0'].unique()
print df_train['num_var13_medio'].unique()
print df_train['num_var13_medio_0'].sum()
print df_train['num_var13_medio'].sum()

[0 3]
[0 3]
6
6


In [ ]:
# Data Cleaning - Drop duplicate columns
df_train.T.drop_duplicates().T

In [ ]:
print df_train.shape()

In [51]:
from collections import OrderedDict

df_train_unique = []
for col in df_train:
    # for col in df_train:
    #     unique_count = len(df_train[col].unique())
    #     df_train_unique[col] = unique_count


In [44]:
df_train_unique

{'ID': 76020,
 'TARGET': 2,
 'delta_imp_amort_var18_1y3': 2,
 'delta_imp_amort_var34_1y3': 2,
 'delta_imp_aport_var13_1y3': 27,
 'delta_imp_aport_var17_1y3': 7,
 'delta_imp_aport_var33_1y3': 9,
 'delta_imp_compra_var44_1y3': 17,
 'delta_imp_reemb_var13_1y3': 2,
 'delta_imp_reemb_var17_1y3': 3,
 'delta_imp_reemb_var33_1y3': 2,
 'delta_imp_trasp_var17_in_1y3': 3,
 'delta_imp_trasp_var17_out_1y3': 2,
 'delta_imp_trasp_var33_in_1y3': 3,
 'delta_imp_trasp_var33_out_1y3': 2,
 'delta_imp_venta_var44_1y3': 5,
 'delta_num_aport_var13_1y3': 6,
 'delta_num_aport_var17_1y3': 6,
 'delta_num_aport_var33_1y3': 4,
 'delta_num_compra_var44_1y3': 9,
 'delta_num_reemb_var13_1y3': 2,
 'delta_num_reemb_var17_1y3': 3,
 'delta_num_reemb_var33_1y3': 2,
 'delta_num_trasp_var17_in_1y3': 3,
 'delta_num_trasp_var17_out_1y3': 2,
 'delta_num_trasp_var33_in_1y3': 3,
 'delta_num_trasp_var33_out_1y3': 2,
 'delta_num_venta_var44_1y3': 5,
 'imp_amort_var18_hace3': 1,
 'imp_amort_var18_ult1': 3,
 'imp_amort_var34_hace3':

### Data Cleaning


In [7]:
df_train.columns

Index([u'ID', u'var3', u'var15', u'imp_ent_var16_ult1',
       u'imp_op_var39_comer_ult1', u'imp_op_var39_comer_ult3',
       u'imp_op_var40_comer_ult1', u'imp_op_var40_comer_ult3',
       u'imp_op_var40_efect_ult1', u'imp_op_var40_efect_ult3',
       u'imp_op_var40_ult1', u'imp_op_var41_comer_ult1',
       u'imp_op_var41_comer_ult3', u'imp_op_var41_efect_ult1',
       u'imp_op_var41_efect_ult3', u'imp_op_var41_ult1',
       u'imp_op_var39_efect_ult1', u'imp_op_var39_efect_ult3',
       u'imp_op_var39_ult1', u'imp_sal_var16_ult1', u'ind_var1_0', u'ind_var1',
       u'ind_var2_0', u'ind_var2', u'ind_var5_0', u'ind_var5', u'ind_var6_0',
       u'ind_var6', u'ind_var8_0', u'ind_var8', u'ind_var12_0', u'ind_var12',
       u'ind_var13_0', u'ind_var13_corto_0', u'ind_var13_corto',
       u'ind_var13_largo_0', u'ind_var13_largo', u'ind_var13_medio_0',
       u'ind_var13_medio', u'ind_var13', u'ind_var14_0', u'ind_var14',
       u'ind_var17_0', u'ind_var17', u'ind_var18_0', u'ind_var18',
     

In [45]:
#df_train.select_dtypes(include=['int64'])

### Algorithm Approach

With an imbalanced dataset, we should try doing the following:

* Random oversampling (add instances) to the underrepresented TARGET type (1)
* Random undersampling (remove instances) from the overrepresented TARGET type (0)
* change our algorithm to overweight the underrepresented TARGET type (1)
* Definitely use AUC (Area Under the ROC Curve) to get the real performance (or else you'll get misleading results)
* We should think about algorithms like AdaBoost, Decision Trees, or some type of ensemble 

### TODO: Plot the Results

Confusion Matrix will show the positives, negatives, false positives

## Split the Data

I mentioned that we should only glimpse at our data because we don't want to cheat and look at all our data (which would cause our algorithm to overfit). So before we do any in-depth analysis, we want to split our data into train and test with __X__ being the features we might want and __y__ being the target.

In [8]:
X = df_train.drop(['TARGET'], axis=1)  # We want all the data except the Target
y = df_train.TARGET  # We only want the target

In [11]:
X.info()
X.head()  # Let's double check the column TARGET is no longer there

<class 'pandas.core.frame.DataFrame'>
Int64Index: 76020 entries, 0 to 76019
Columns: 370 entries, ID to var38
dtypes: float64(111), int64(259)
memory usage: 215.2 MB


In [ ]:
y.head()  # let's double check the column TARGET is the only data there

### Variations to splitting the data and validating

There's a few ways we can split and validate our data. First, we can do a simple __train and test split__ (default .75 to .25); the advantage with this method is that it's quick and easy, but the issue is we don't use all the training data. Another method is to do __K-Fold__ where you split your data K times, then do validation and average up the scores; the advantage with this method is you use more data, but this takes more computing time. Depending on how balanced your dataset, you might also want to consider __StratifiedKFold__.

In [ ]:
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y)
X_train.shape, X_test.shape

In [ ]:
from sklearn.cross_validation import KFold
